### PR TITLE
Generalise Sinsemilla API

### PR DESCRIPTION
**Pull request by therealyingtong**
Please review with whitespace changes hidden.

### CODE REVIEW

File: src/primitives/sinsemilla.rs [hunk 1]
```diff
@@ -64,59 +61,87 @@ impl<I: Iterator<Item = bool>> Iterator for Pad<I> {
     }
 }
 
-#[allow(non_snake_case)]
-fn Q(domain_prefix: &str) -> pallas::Point {
-    pallas::Point::hash_to_curve(GROUP_HASH_Q)(domain_prefix.as_bytes())
+pub trait HashDomain {
```
Comment (str4d) [inline src/primitives/sinsemilla.rs:64]:
```suggestion
/// Trait representing a domain in which [`sinsemilla::hash_to_point`] and
/// [`sinsemilla::hash`] can be used.
///
/// [`sinsemilla::hash_to_point`]: self::hash_to_point
/// [`sinsemilla::hash`]: self::hash
pub trait HashDomain {
```

File: src/primitives/sinsemilla.rs [hunk 2]
```diff
@@ -64,59 +61,87 @@ impl<I: Iterator<Item = bool>> Iterator for Pad<I> {
...
+    #[allow(non_snake_case)]
+    fn Q(&self) -> pallas::Point;
+}
+
+pub struct Domain(pub &'static str);
```
Comment (str4d) [inline src/primitives/sinsemilla.rs:69]:
```suggestion
/// A domain with a specific prefix string, in which [`sinsemilla::hash_to_point`] and
/// [`sinsemilla::hash`] can be used.
///
/// [`sinsemilla::hash_to_point`]: self::hash_to_point
/// [`sinsemilla::hash`]: self::hash
pub struct Domain(pub &'static str);
```

File: src/primitives/sinsemilla.rs [hunk 3]
```diff
@@ -64,59 +61,87 @@ impl<I: Iterator<Item = bool>> Iterator for Pad<I> {
...
 /// `SinsemillaHash` from [§ 5.4.1.9][concretesinsemillahash].
 ///
 /// [concretesinsemillahash]: https://zips.z.cash/protocol/nu5.pdf#concretesinsemillahash
-pub(crate) fn hash(domain_prefix: &str, msg: impl Iterator<Item = bool>) -> pallas::Base {
-    extract_p(&hash_to_point(domain_prefix, msg))
+pub(crate) fn hash<D: HashDomain>(domain: &D, msg: impl Iterator<Item = bool>) -> pallas::Base {
+    extract_p(&hash_to_point(domain, msg))
+}
+
+pub trait CommitDomain: HashDomain {
```
Comment (str4d) [inline src/primitives/sinsemilla.rs:102]:
```suggestion
/// Trait representing a domain in which [`sinsemilla::commit`] and
/// [`sinsemilla::short_commit`] can be used.
///
/// [`sinsemilla::commit`]: self::commit
/// [`sinsemilla::short_commit`]: self::short_commit
pub trait CommitDomain: HashDomain {
```

File: src/primitives/sinsemilla.rs [hunk 4]
```diff
@@ -64,59 +61,87 @@ impl<I: Iterator<Item = bool>> Iterator for Pad<I> {
...
+    #[allow(non_snake_case)]
+    fn R(&self) -> pallas::Point;
+}
+
+pub struct Comm(pub &'static str);
```
Comment (str4d) [inline src/primitives/sinsemilla.rs:107]:
```suggestion
/// A domain with a specific prefix string, in which [`sinsemilla::commit`] and
/// [`sinsemilla::short_commit`] can be used.
///
/// [`sinsemilla::commit`]: self::commit
/// [`sinsemilla::short_commit`]: self::short_commit
pub struct Comm(pub &'static str);
```

**Review by str4d — APPROVED** [review]
utACK modulo comments, but I co-authored this PR so this still needs two reviews.

File: src/primitives/sinsemilla.rs [hunk 5]
```diff
@@ -64,59 +61,107 @@ impl<I: Iterator<Item = bool>> Iterator for Pad<I> {
...
+/// A domain with a specific prefix string, in which [`sinsemilla::hash_to_point`] and
+/// [`sinsemilla::hash`] can be used.
+///
+/// [`sinsemilla::hash_to_point`]: self::hash_to_point
+/// [`sinsemilla::hash`]: self::hash
+pub struct Domain(pub &'static str);
+
+impl HashDomain for Domain {
+    fn Q(&self) -> pallas::Point {
+        pallas::Point::hash_to_curve(Q_PERSONALIZATION)(self.0.as_bytes())
```
Comment (daira) [inline src/primitives/sinsemilla.rs:83]: I suggest doing the `hash_to_curve` only when constructing a domain, not later (we only need to access Q on a domain, not the original string).

File: src/primitives/sinsemilla.rs [hunk 6]
```diff
@@ -64,59 +61,107 @@ impl<I: Iterator<Item = bool>> Iterator for Pad<I> {
...
     let padded: Vec<_> = Pad::new(msg).collect();
 
-    let hasher_S = pallas::Point::hash_to_curve(GROUP_HASH_S);
+    let hasher_S = pallas::Point::hash_to_curve(S_PERSONALIZATION);
     let S = |chunk: &[bool]| hasher_S(&lebs2ip_k(chunk).to_le_bytes());
 
     padded
         .chunks(K)
-        .fold(Q(domain_prefix), |acc, chunk| acc.double() + S(chunk))
+        .fold(domain.Q(), |acc, chunk| acc.double() + S(chunk))
```
Comment (daira) [inline src/primitives/sinsemilla.rs:102]:
```suggestion
        .fold(domain.0, |acc, chunk| acc.double() + S(chunk))
```

File: src/primitives/sinsemilla.rs [hunk 7]
```diff
@@ -64,59 +61,107 @@ impl<I: Iterator<Item = bool>> Iterator for Pad<I> {
...
+    #[allow(non_snake_case)]
+    fn Q(&self) -> pallas::Point;
+}
+
+/// A domain with a specific prefix string, in which [`sinsemilla::hash_to_point`] and
+/// [`sinsemilla::hash`] can be used.
+///
+/// [`sinsemilla::hash_to_point`]: self::hash_to_point
+/// [`sinsemilla::hash`]: self::hash
+pub struct Domain(pub &'static str);
```
Comment (daira) [inline src/primitives/sinsemilla.rs:79]:
```suggestion
pub struct Domain(pub &'static pallas::Point);
```

File: src/primitives/sinsemilla.rs [hunk 8]
```diff
@@ -64,59 +61,107 @@ impl<I: Iterator<Item = bool>> Iterator for Pad<I> {
...
+
+impl HashDomain for Domain {
+    fn Q(&self) -> pallas::Point {
+        pallas::Point::hash_to_curve(Q_PERSONALIZATION)(self.0.as_bytes())
+    }
```
Comment (daira) [inline src/primitives/sinsemilla.rs:84]:
```suggestion
impl Domain {
    fn new(domain: &str) -> Self {
        Domain(pallas::Point::hash_to_curve(Q_PERSONALIZATION)(domain.as_bytes()))
    }
```

File: src/primitives/sinsemilla.rs [hunk 9]
```diff
@@ -64,59 +61,107 @@ impl<I: Iterator<Item = bool>> Iterator for Pad<I> {
...
+
+/// Trait representing a domain in which [`sinsemilla::commit`] and
+/// [`sinsemilla::short_commit`] can be used.
+///
+/// [`sinsemilla::commit`]: self::commit
+/// [`sinsemilla::short_commit`]: self::short_commit
+pub trait CommitDomain: HashDomain {
+    #[allow(non_snake_case)]
+    fn R(&self) -> pallas::Point;
+}
```
Comment (daira) [inline src/primitives/sinsemilla.rs:120]:
```suggestion
pub struct CommitDomain {
    #[allow(non_snake_case)]
    Q: pallas::Point,
    #[allow(non_snake_case)]
    R: pallas::Point,
}
```

File: src/primitives/sinsemilla.rs [hunk 10]
```diff
@@ -64,59 +61,107 @@ impl<I: Iterator<Item = bool>> Iterator for Pad<I> {
...
+    }
+}
+
+impl CommitDomain for Comm {
+    fn R(&self) -> pallas::Point {
+        let r_prefix = self.0.to_owned() + "-r";
+        let hasher_r = pallas::Point::hash_to_curve(&r_prefix);
+        hasher_r(&[])
+    }
 }
```
Comment (daira) [inline src/primitives/sinsemilla.rs:142]:
```suggestion
impl CommitDomain {
    fn new(domain: &str) -> Self {
        let m_prefix = domain.to_owned() + "-M";
        let r_prefix = domain.to_owned() + "-r";
        CommitDomain {
            Q: pallas::Point::hash_to_curve(Q_PERSONALIZATION)(m_prefix.as_bytes()),
            R: pallas::Point::hash_to_curve(r_prefix.as_bytes())(&[]),
        }
    }
}
```

File: src/primitives/sinsemilla.rs [hunk 11]
```diff
@@ -64,59 +61,107 @@ impl<I: Iterator<Item = bool>> Iterator for Pad<I> {
...
     msg: impl Iterator<Item = bool>,
     r: &pallas::Scalar,
 ) -> pallas::Point {
-    let m_prefix = domain_prefix.to_owned() + "-M";
-    let r_prefix = domain_prefix.to_owned() + "-r";
-
-    let hasher_r = pallas::Point::hash_to_curve(&r_prefix);
-
-    hash_to_point(&m_prefix, msg) + hasher_r(&[]) * r
+    hash_to_point(domain, msg) + domain.R() * r
```
Comment (daira) [inline src/primitives/sinsemilla.rs:153]:
```suggestion
    hash_to_point(domain, msg) + domain.R * r
```

File: src/primitives/sinsemilla.rs [hunk 12]
```diff
@@ -64,59 +61,107 @@ impl<I: Iterator<Item = bool>> Iterator for Pad<I> {
...
 }
 
 /// `SinsemillaCommit` from [§ 5.4.7.4][concretesinsemillacommit].
 ///
 /// [concretesinsemillacommit]: https://zips.z.cash/protocol/nu5.pdf#concretesinsemillacommit
 #[allow(non_snake_case)]
-pub(crate) fn commit(
-    domain_prefix: &str,
+pub(crate) fn commit<D: CommitDomain>(
+    domain: &D,
```
Comment (daira) [inline src/primitives/sinsemilla.rs:149]:
```suggestion
pub(crate) fn commit(
    domain: &CommitDomain,
```

File: src/primitives/sinsemilla.rs (hunk 10 above)
Comment (str4d) [reply src/primitives/sinsemilla.rs:142]: I'd tweak the above to `Q: HashDomain::new(m_prefix.as_bytes()).Q` which better reflects the relationship.

File: src/primitives/sinsemilla.rs [hunk 13]
```diff
@@ -64,59 +61,124 @@ impl<I: Iterator<Item = bool>> Iterator for Pad<I> {
...
-    msg: impl Iterator<Item = bool>,
-    r: &pallas::Scalar,
-) -> pallas::Base {
-    extract_p(&commit(domain_prefix, msg, r))
+        let hasher_S = pallas::Point::hash_to_curve(S_PERSONALIZATION);
+        let S = |chunk: &[bool]| hasher_S(&lebs2ip_k(chunk).to_le_bytes());
+
+        padded
+            .chunks(K)
+            .fold(self.Q, |acc, chunk| acc.double() + S(chunk))
```
Comment (str4d) [inline src/primitives/sinsemilla.rs:145]: Hmm, I do not like this duplication. Instead, store the `HashDomain` you are using to construct `Q`, and use that directly.

File: src/primitives/sinsemilla.rs [hunk 14]
```diff
@@ -64,59 +61,124 @@ impl<I: Iterator<Item = bool>> Iterator for Pad<I> {
...
+}
 
-    let hasher_r = pallas::Point::hash_to_curve(&r_prefix);
+impl CommitDomain {
+    /// Construct a new CommitDomain with a specific prefix string
+    pub(crate) fn new(domain: &str) -> Self {
+        let m_prefix = domain.to_owned() + "-M";
+        let r_prefix = domain.to_owned() + "-r";
+        CommitDomain {
+            Q: HashDomain::new(&m_prefix).Q,
```
Comment (str4d) [inline src/primitives/sinsemilla.rs:128]:
```suggestion
            M: HashDomain::new(&m_prefix),
```

File: src/primitives/sinsemilla.rs [hunk 15]
```diff
@@ -64,59 +61,124 @@ impl<I: Iterator<Item = bool>> Iterator for Pad<I> {
...
 #[allow(non_snake_case)]
-pub(crate) fn commit(
-    domain_prefix: &str,
-    msg: impl Iterator<Item = bool>,
-    r: &pallas::Scalar,
-) -> pallas::Point {
-    let m_prefix = domain_prefix.to_owned() + "-M";
-    let r_prefix = domain_prefix.to_owned() + "-r";
+pub struct CommitDomain {
+    Q: pallas::Point,
```
Comment (str4d) [inline src/primitives/sinsemilla.rs:118]:
```suggestion
    M: HashDomain,
```

File: src/primitives/sinsemilla.rs [hunk 16]
```diff
@@ -64,59 +61,124 @@ impl<I: Iterator<Item = bool>> Iterator for Pad<I> {
...
+        &self,
+        msg: impl Iterator<Item = bool>,
+        r: &pallas::Scalar,
+    ) -> pallas::Base {
+        extract_p(&self.commit(msg, r))
+    }
+
+    /// Return `Q`
+    #[allow(non_snake_case)]
+    pub(crate) fn Q(&self) -> pallas::Point {
```
Comment (str4d) [inline src/primitives/sinsemilla.rs:173]: Where will this be used? (Same question for the other unused accessors).

**Review by str4d — CHANGES_REQUESTED** [review]
The documentation links all need fixing now that the standalone methods have been moved onto domain structs.

File: src/primitives/sinsemilla.rs (hunk 16 above)
Comment (therealyingtong) [reply src/primitives/sinsemilla.rs]: I'm using `R()` in tests for Orchard constants:
- https://github.com/zcash/orchard/pull/43/files#diff-67648b0a12585bf10d8511f896af5c7226e18f770cf74cfb8aace4af187766aaR43
- https://github.com/zcash/orchard/pull/43/files#diff-0236c6fdfb7dc339abd6f277fab892bc8a17a158107dd164a5ba7e330348fdb8R43

Comment (str4d) [reply src/primitives/sinsemilla.rs]: If these are only used in tests, then place them behind `#[cfg(test)]`.

**Review by str4d — APPROVED** [review]
utACK

**Pull request merged**